Add time-dependent $B_s^0 \rightarrow \phi \mu^+\mu^-$ observables

schmitse opened the pull request:
This PR adds angular observables measurable in a time-dependent angular analysis of $B_s^0 \rightarrow \phi \mu^+\mu^-$ decays and updates the observables defined for the time-integrated case. 
In general the implementation follows Virto et al in arXiv 1502.05509, where the methodology of $B \rightarrow V \ell\ell$ are explored when the decay products are $CP$-eigenstates. 

The main differences concern the normalisation of the angular coefficients, the integrated decay rate in a given $q^2$ region and the definition of the observables themselves. 
In general this PR introduces six kind of angular coefficients: 
  - $K$, $W$: which are equivalent to $S$ and $A$ in the $B^0 \rightarrow K^{\ast 0}\mu^+\mu^-$ decays
  -  $\mathcal{H}$, $\mathcal{Z}$: which are added through the time-dependent $CP$-V
  - $S$, $A$: which are measured in the time-integrated case for $B_s^0 \rightarrow \phi\mu^+\mu^-$ (The name difference with respect to $K^{\ast 0}$ is unfortunate but induced by the fact that the time-integrated LHCb measured used this nomenclature) 

Key changes comprise: 
  - `BsToPhiDilepton`: 
    - Add two new public functions to retrieve the $J$ coefficients both differential and integrated. 
    - Change the prefactors for the $J_i$ to be aligned with Virto et al in arXiv 1502.05509, this results in a different prefactor of the decay rate as well.  
  - `BsToPhiDileptonAndConjugate`: 
    - Add new struct for the $s_i$ coefficients `AngularsCoefficients`
    - Add calculation of the $s_i$ following Virto et al in arXiv 1502.05509
    - Add functions to calculate the angular coefficients following Virto et al in arXiv arXiv 1502.05509
    - Add `decay_rate` function following Virto et al in arXiv 1502.05509
  - `obervables.cc`: Add definitions for all of the angular coefficients in their "theory" and "LHCb" conventions. I prepared some macros to reduce manual writing in the future but they are currently not used. 
  - `hadron-properties`: Update the phase of $q/p$ for the $B_s^0$ system to use the most recent value from HFAG. 

I will prepare a small accompanying pdf document that details the changes in the code, why i think they are necessary and adds comparisons to the implementation in flavio.

schmitse committed:
[rare-b-decays] add new Bs -> phi observables

B_s->phi_ll::Gamma uses the implementation of BsToPhiDileptonAndConjugate
reflecting on Virto et al in 1502.05509.
Add all K,W,H,Z,S, and A observables in integrated and differential as
well as "theory" and "LHCb" convention.
Change between "theory" and "LHCb" is a minus sign for coefficients
4, 6s, 6c, 7, and 9, see Figure 4 in 1506.03970.

schmitse committed:
[parameters] Update B_s::q_over_p_phase from HFAG

schmitse commented:
Hi Méril, 

sorry i did not see your first response before. 
I agree discussing tomorrow in person will be more efficient. 
The reason i was switching to use this definition is because then i can use all numerical factors from 1502.05509 for the $\phi$ implementation. 
In general i agree that having a streamlined version is desirable but i did not have the time to make sure i am not producing more typos than correct results. 
Using the parameters for $x$, $y$, $\Gamma$ makes sense. I was trying to get a working version to compare against flavio and did not think about best practices for the details.

mreboud commented on eos/rare-b-decays/bs-to-phi-ll.cc:
```diff
@@ -91,7 +91,7 @@ namespace eos
             Context ctx("When constructing Bs->Phill observables");
 
-            std::string tag = o.get("tag"_ok, "");
+            std::string tag = o.get("tag"_ok, "GvDV2020");
```
Please keep this default value empty. We don't want to provide a default value here.

mreboud commented on eos/rare-b-decays/bs-to-phi-ll.cc:
```diff
@@ -118,7 +118,7 @@ namespace eos
         }
 
-        inline std::array<double, 12> angular_coefficients_array(const BsToPhiDilepton::Amplitudes & A, const double & s) const
+        std::array<double, 12> angular_coefficients_array(const BsToPhiDilepton::Amplitudes & A, const double & s) const
```
Can you put the `inline` statement back here?

mreboud commented on eos/rare-b-decays/bs-to-phi-ll.cc:
```diff
@@ -130,7 +130,7 @@ namespace eos
 
             // j1s
-            result[0] = 3.0 / 4.0 * (
+            result[0] = (
```
As discussed, we would like to keep the current definition for the j's observables. I can take care of rearranging the factors of 3/4 once the rest is ready, if you want.

mreboud commented on eos/rare-b-decays/bs-to-phi-ll.cc:
```diff
@@ -241,7 +241,7 @@ namespace eos
         {
             // cf. [BHvD2010], p. 6, eq. (2.7)
-            return 2.0 * a_c.j1s + a_c.j1c - 1.0 / 3.0 * (2.0 * a_c.j2s + a_c.j2c);
+            return 2.0 * a_c.j1s + a_c.j1c - 1.0 / 4.0 * (2.0 * a_c.j2s + a_c.j2c);
```
A factor of  3/4 is missing in the first two terms

mreboud commented on eos/rare-b-decays/bs-to-phi-ll.cc:
```diff
@@ -913,8 +924,9 @@ The azimuthal angle between the Kbar-K plane and the l^+l^- plane.";
-        bstophidilepton_conjugate(parameters, options + Options{{"cp-conjugate"_ok, "false"}})
+        bstophidilepton(parameters, options + Options{{"cp-conjugate"_ok, "false"}}),
+        bstophidilepton_conjugate(parameters, options + Options{{"cp-conjugate"_ok, "true"}}),
+        m_y(0.129), m_x(26.79), m_gamma_s(0.6578428049823233)
```
Can you use parameters instead of class variables?

mreboud commented on eos/rare-b-decays/bs-to-phi-ll.cc:
```diff
@@ -970,10 +981,10 @@ The azimuthal angle between the Kbar-K plane and the l^+l^- plane.";
         double beta = std::sqrt(beta2);
 
-        const complex<double> expiphi(cos(phiBs), sin(phiBs));
+        const complex<double> expiphi(cos(phiBs), -1.0 * sin(phiBs)); // e^iphi is the phase of -q/p
```
I am not so sure, can you check that?

mreboud commented on eos/rare-b-decays/bs-to-phi-ll.cc:
```diff
@@ -1095,35 +1106,562 @@ The azimuthal angle between the Kbar-K plane and the l^+l^- plane.";
 
     double
-    BsToPhiDileptonAndConjugate::integrated_H_1c(const double & s_min, const double & s_max) const
+    BsToPhiDileptonAndConjugate::integrated_H(const double & s_min, const double & s_max, const std::string & name) const
```
We need to discuss this in more details :)

schmitse replied:
In the most recent version i have one suggestion using an enum instead of a string to curry the function instead. (See the integrated and differential S functions).

mreboud reviewed (changes requested):
Thanks a lot for the PR @DominikSue. Here is a first round of comments. I think we need to discuss with @dvandyk how we want to implement some of the functions you suggested.

schmitse committed:
[rare-b-decays] Revert 3/4 in J_i and make angular coefficients inline function

Revert J_i to use factor 3/4 for example from 1212.2321
and adjust the derived equations
Make `angular_coefficients_array` inline to allow for more optimisations again
Use enum to curry integrated_S and differential_S

schmitse committed:
[rare-b-decays] update Bs -> phi mumu based on arXiv 1502.05509

BsToPhiDilepton: Add two public functions to retrieve the
integrated and differential angular coefficients.
Adjusted the prefactors for the J_i to use the convention from
Virto et al in 1502.05509.
This also means that some pre-factors in functions change slightly.

BsToPhiDileptonAndConjugate:
Add new properties m_y, m_x, m_gamma_s corresponding to the
DeltaGamma/Gamma_s, DeltaM/Gamma_s and Gamma_s.
Add new struct: AngularsCoefficients mirrors AngularhCoefficients.
Add calculation of the s_i coefficients.

Modify AngularhCoefficients to use the same prefactors as in
Virto et al 1502.05509.
Add functions for the integrated and differential angular
coefficients S, A, K, W, H, and Z.
The K and W correspond to the S and A in the Kst case and
used to be called S and A for the phi case as well.
The new S and A are based on Virto et al 1502.05509 and reflect
on the time-integration needed to be performed which introduces
factors of y / x and h / s.
The definition of the functions is handled using macros
to avoid typos and repetetive typing.

schmitse committed:
[rare-b-decays] fix typo in h1c for BsToPhiLL

schmitse commented:
I should have fixed the commit messages, split the commit where the typo is fixed from the others adding new things and reverted the inline and tag change i had made locally.

schmitse commented:
Also i re-added the factor $3/4$ in the $J_i$ definitions and accordingly in the angular observable functions. 
For the $h_i$ and $s_i$ i don't have any different factors compared to Virto in 1502.05509.

schmitse committed:
[rare-b-decays] add optimised angular observables for Bs->phill

schmitse commented:
Hi Méril, 
we have not had a meeting yet, i will follow up with Danny offline.
